node/narwhal/src/helpers/cache.rs

```diff
@@ -131,47 +131,76 @@ mod tests {
+            IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1))
+        }
+    }
 
-        // Check that the cache contains the solution.
-        assert_eq!(cache.seen_inbound_certificates.read().len(), 1);
+    impl Input for SocketAddr {
+        fn input() -> Self {
+            SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 1234)
+        }
+    }
 
-        // Insert the same solution again.
-        assert_eq!(cache.insert_inbound_certificate(certificate_id, 5), 2);
+    impl Input for Field<CurrentNetwork> {
+        fn input() -> Self {
+            Field::from_u8(1)
+        }
+    }
 
-        // Check that the cache still contains the solution.
-        assert_eq!(cache.seen_inbound_certificates.read().len(), 1);
+    impl Input for TransmissionID<CurrentNetwork> {
+        fn input() -> Self {
+            TransmissionID::Transaction(Default::default())
+        }
     }
 
-    #[test]
-    fn test_inbound_transmission() {
-        let cache = Cache::<CurrentNetwork>::default();
-        let transmission = TransmissionID::Transaction(Default::default());
+    const INTERVAL_IN_SECS: i64 = 5;
+
+    macro_rules! test_cache_fields {
+        ($($name:ident),*) => {
+            $(
+                paste::paste! {
+                    #[test]
+                    fn [<test_seen_ $name s>]() {
+                        let cache = Cache::<CurrentNetwork>::default();
+                        let input = Input::input();
 
-        // Check that the cache is empty.
-        assert_eq!(cache.seen_inbound_transmissions.read().len(), 0);
+                        // Check that the cache is empty.
+                        assert!(cache.[<seen_ $name s>].read().is_empty());
 
-        // Insert a transmission.
-        assert_eq!(cache.insert_inbound_transmission(transmission, 5), 1);
+                        // Insert an input, recent events should be 1.
+                        assert_eq!(cache.[<insert_ $name>](input, INTERVAL_IN_SECS), 1);
 
-        // Check that the cache contains the transmission.
-        assert_eq!(cache.seen_inbound_transmissions.read().len(), 1);
+                        // Check that the cache contains the input.
+                        assert_eq!(cache.[<seen_ $name s>].read().len(), 1);
 
-        // Insert the same transmission again.
-        assert_eq!(cache.insert_inbound_transmission(transmission, 5), 2);
+                        // Insert the same input again, recent events should be 2.
+                        assert_eq!(cache.[<insert_ $name>](input, INTERVAL_IN_SECS), 2);
+
+                        // Check that the cache still contains the input.
+                        assert_eq!(cache.[<seen_ $name s>].read().len(), 1);
+                    }
+                }
+            )*
+        }
+    }
 
-        // Check that the cache still contains the transmission.
-        assert_eq!(cache.seen_inbound_transmissions.read().len(), 1);
+    test_cache_fields! {
+       inbound_connection,
```

Comment:
Note: using paste to construct the macro means the common name subset (field + insert func) needs to be passed in here as opposed to the full field name.